### PR TITLE
Refactors to the code 

### PR DESCRIPTION
**Pull request by Pirulax**
Alright, here I'll try to refactor a lot of code, while in the same time having meaningful commits.
Im creating this PR, so the changes can be reviewed if needed.
Also, Im trying not to refactor the code too much, so I dont introduce bugs, because that woudnt be fun.

Mostly I refactor loop to use 'range based for loop' I also add a lot of 'auto', and 'constexpr' where needed.

### CODE REVIEW

**Comment by Pirulax** [comment]
Yeah so I've already fucked up my clean commit list... 😁😁

File: Client/core/Graphics/CRenderItemManager.cpp [hunk 1]
```diff
@@ -328,8 +327,7 @@ void CRenderItemManager::NotifyContructRenderItem(CRenderItem* pItem)
 ////////////////////////////////////////////////////////////////
 void CRenderItemManager::NotifyDestructRenderItem(CRenderItem* pItem)
 {
-    assert(MapContains(m_CreatedItemList, pItem));
-    MapRemove(m_CreatedItemList, pItem);
+    assert(!m_CreatedItemList.insert(pItem).second); // assert if item was already in the collection.
```
Comment (ccw808) [inline Client/core/Graphics/CRenderItemManager.cpp]: Should be remove

File: Client/core/Graphics/CGraphics.cpp [hunk 2]
```diff
@@ -668,18 +669,10 @@ float CGraphics::GetDXTextExtentW(const wchar_t* wszText, float fScale, LPD3DXFO
...
-        {
-            const wchar_t c = wszText[i];
-
-            if (c == L' ')
-                ++spaceCount;
-            else
-                break;
-        }
+        size_t       spaceCount = 0;
+        for (int i = textLength - 1; i >= 0 && wszText[i] == L' '; --i) { spaceCount++; } // count spaces starting from the end
```
Comment (ccw808) [inline Client/core/Graphics/CGraphics.cpp]: Does this line follow our formatting rules?

File: Client/core/Graphics/CRenderItemManager.cpp (hunk 1 above)
Comment (Pirulax) [reply Client/core/Graphics/CRenderItemManager.cpp]: Nice to see you.
Do you mean the assert?

edit: nvm, its a typo, I see. (I thought you meant 'removed' at the first glance, then realized what you really meant.

File: Client/core/Graphics/CGraphics.cpp (hunk 2 above)
Comment (Pirulax) [reply Client/core/Graphics/CGraphics.cpp]: I wasnt sure, but it looked neater like this.

File: Client/core/Graphics/CRenderItemManager.cpp (hunk 1 above)
Comment (Pirulax) [reply Client/core/Graphics/CRenderItemManager.cpp]: Fixed.

File: Client/core/Graphics/CGraphics.cpp (hunk 2 above)
Comment (Lpsd) [reply Client/core/Graphics/CGraphics.cpp]: It should be
```cpp
for (int i = textLength - 1; i >= 0 && wszText[i] == L' '; --i)
    spaceCount++;
```

File: Client/core/Graphics/CGraphics.cpp [hunk 3]
```diff
@@ -2265,36 +2264,40 @@ void CGraphics::DrawProgressMessage(bool bPreserveBackbuffer)
...
         if (FAILED(D3DXLoadSurfaceFromSurface(pNewSurface, NULL, NULL, pCurrentSurface, NULL, NULL, D3DX_FILTER_TRIANGLE, 0)))
-            break;
+        goto cleanup_and_return;
 
         // Data out
         if (!CopyDataFromSurface(pNewSurface, outBuffer))
-            break;
+        goto cleanup_and_return;
 
         bResult = true;
```
Comment (qaisjp) [inline Client/core/Graphics/CGraphics.cpp]: Indentation problems.

File: Shared/sdk/CVector4D.h [hunk 4]
```diff
@@ -20,28 +20,28 @@
 class CVector4D
 {
 public:
-    CVector4D()
+    constexpr CVector4D() :
+        fX(0),
+        fY(0),
+        fZ(0),
+        fW(0)
```
Comment (qaisjp) [inline Shared/sdk/CVector4D.h]: this can just be `int fX = 0;` in member declaration

Comment (Pirulax) [reply Shared/sdk/CVector4D.h]: actually, you're right. I just wasnt sure if I should do it that way.

File: Client/core/Graphics/CGraphics.cpp [hunk 5]
```diff
@@ -158,46 +158,47 @@ void CGraphics::DrawString(int iX, int iY, unsigned long dwColor, float fScale,
 // Slow
 void CGraphics::DrawStringOutline(const RECT& rect, unsigned long ulColor, const wchar_t* szText, unsigned long ulFormat, LPD3DXFONT pDXFont)
 {
-    const uint   uiKernelSizeX = 5;
-    const uint   uiKernelSizeY = 5;
-    const float* pKernel;
+    constexpr int iKernelSizeX = 5;
+    constexpr int iKernelSizeY = 5;
```
Comment (qaisjp) [inline Client/core/Graphics/CGraphics.cpp:162]: why has the type changed from uint to int?

Comment (Pirulax) [reply Client/core/Graphics/CGraphics.cpp:162]: it was unnecessary for it to be `uint`, it was converted to `int` anyways in the loop(line 200).
im pretty sure the compiler optimized it nevertheless.

Comment (qaisjp) [reply Client/core/Graphics/CGraphics.cpp:162]: I see, so iKernelSizeX is eventually used to create a RECT, which uses `long`, which is essentially just a (signed) int. Okay 👍

Comment (Pirulax) [reply Client/core/Graphics/CGraphics.cpp:162]: Exactly. The compiler probably noticed it already, but i was like why not to change it.

File: Client/core/Graphics/CGraphics.cpp [hunk 6]
```diff
@@ -483,43 +485,42 @@ void CGraphics::EndDrawBatch()
 //
 void CGraphics::CheckModes(EDrawModeType newDrawMode, EBlendModeType newBlendMode)
 {
-    bool bDrawModeChanging = (m_CurDrawMode != newDrawMode);
-    bool bBlendModeChanging = (m_CurBlendMode != newBlendMode);
-    // Draw mode changing?
-    if (bDrawModeChanging || bBlendModeChanging)
```
Comment (qaisjp) [inline Client/core/Graphics/CGraphics.cpp:489]: I think this old code is clearer and makes it more obvious what's happening

Comment (Pirulax) [reply Client/core/Graphics/CGraphics.cpp:489]: Yeah, you're right. I'll change it back.

File: Client/core/Graphics/CGraphics.cpp [hunk 7]
```diff
@@ -27,7 +27,7 @@ using namespace std;
 template <>
 CGraphics* CSingleton<CGraphics>::m_pSingleton = NULL;
 
-const SColor g_rectEdgePixelsData[] = {
+static const SColor g_rectEdgePixelsData[] = {
```
Comment (qaisjp) [inline Client/core/Graphics/CGraphics.cpp]: Isn't this in the data section already? Why does it need a static?

File: Client/core/Graphics/CGraphics.cpp [hunk 8]
```diff
@@ -483,50 +485,59 @@ void CGraphics::EndDrawBatch()
...
-            m_pPrimitiveBatcher->Flush();
-        }
-        else if (m_CurDrawMode == EDrawMode::PRIMITIVE_MATERIAL)
-        {
-            m_pPrimitiveMaterialBatcher->Flush();
-        }
+        // Draw mode changing?
+        if (!bDrawModeChanging && !bBlendModeChanging)
+            return;
+    }
```
Comment (qaisjp) [inline Client/core/Graphics/CGraphics.cpp:495]: What is the extra scope here for? Also there's not really a need for const here

File: Client/core/Graphics/CGraphics.cpp [hunk 9]
```diff
@@ -483,50 +485,59 @@ void CGraphics::EndDrawBatch()
...
+    switch (newDrawMode)
+    {
+    case EDrawMode::DX_SPRITE:
+        m_pDXSprite->Begin(D3DXSPRITE_DONOTSAVESTATE | D3DXSPRITE_DONOTMODIFY_RENDERSTATE);
+        break;
+
+    case EDrawMode::DX_LINE:
+        m_pLineInterface->Begin();
+        break;
+
```
Comment (qaisjp) [inline Client/core/Graphics/CGraphics.cpp]:
```suggestion
```

File: Client/core/Graphics/CGraphics.cpp [hunk 10]
```diff
@@ -483,50 +485,59 @@ void CGraphics::EndDrawBatch()
...
+    }
+
+    SetBlendModeRenderStates(newBlendMode);
+    m_CurBlendMode = newBlendMode;
+    m_CurDrawMode = newDrawMode;
+
```
Comment (qaisjp) [inline Client/core/Graphics/CGraphics.cpp]: unnecessary whitespace

File: Client/core/Graphics/CGraphics.cpp [hunk 11]
```diff
@@ -668,18 +679,11 @@ float CGraphics::GetDXTextExtentW(const wchar_t* wszText, float fScale, LPD3DXFO
...
 
-        for (int i = textLength - 1; i >= 0; --i)
-        {
-            const wchar_t c = wszText[i];
-
-            if (c == L' ')
-                ++spaceCount;
-            else
-                break;
-        }
```
Comment (qaisjp) [inline Client/core/Graphics/CGraphics.cpp]: Old code is clearer.

[No difference in performance.](http://quick-bench.com/FusJ6BGP4JbWPbnZp_03oH5Az98)

<details><summary>Code used, in case quick-bench.com goes down</summary>

```cpp
#include <string.h>

static void Old(benchmark::State& state) {
  // Code inside this loop is measured repeatedly
  for (auto _ : state) {
    const char* x = "hi world!";
    const int textLength = strlen(x);
    int spaceCount = 0;
    for (int i = textLength - 1; i >= 0; --i)	{
      const char c = x[i];
      if (c == ' ')
        ++spaceCount;
      else
        break;
    }
    // Make sure the variable is not optimized away by compiler
    benchmark::DoNotOptimize(x);
    benchmark::DoNotOptimize(spaceCount);
  }
}
// Register the function as a benchmark
BENCHMARK(Old);

static void New(benchmark::State& state) {
  // Code inside this loop is measured repeatedly
  for (auto _ : state) {
    const char* x = "hi world!";
    const int textLength = strlen(x);
    int spaceCount = 0;
    for (int i = textLength - 1; i >= 0 && x[i] == L' '; --i)
      spaceCount++;
    
   
    // Make sure the variable is not optimized away by compiler
    benchmark::DoNotOptimize(x);
    benchmark::DoNotOptimize(spaceCount);
  }
}
BENCHMARK(New);
```

</details>

File: Client/core/Graphics/CGraphics.cpp [hunk 12]
```diff
@@ -739,13 +742,8 @@ int CGraphics::GetTrailingSpacesWidth(ID3DXFont* pDXFont, WString& strText)
 {
     // Count the amount of space characters at the end
     int iSpaceCount = 0;
-    for (auto c = strText.rbegin(); c != strText.rend(); ++c)
-    {
-        if (*c == ' ')
-            ++iSpaceCount;
-        else
-            break;
-    }
```
Comment (qaisjp) [inline Client/core/Graphics/CGraphics.cpp]: Old code is clearer.

File: Client/core/Graphics/CGraphics.cpp [hunk 13]
```diff
@@ -892,10 +891,10 @@ void CGraphics::DrawCircleQueued(float fX, float fY, float fRadius, float fStart
         return;
 
     auto pVecVertices = new std::vector<PrimitiveVertice>();
-    fStartAngle = D3DXToRadian(fStartAngle);
-    fStopAngle = D3DXToRadian(fStopAngle);
```
Comment (qaisjp) [inline Client/core/Graphics/CGraphics.cpp:896]: Possible logical error. fStartAngle is used later.

File: Client/core/Graphics/CRenderItemManager.cpp [hunk 14]
```diff
@@ -312,8 +312,7 @@ CGuiFontItem* CRenderItemManager::CreateGuiFont(const SString& strFullFilePath,
 ////////////////////////////////////////////////////////////////
 void CRenderItemManager::NotifyContructRenderItem(CRenderItem* pItem)
 {
-    assert(!MapContains(m_CreatedItemList, pItem));
-    MapInsert(m_CreatedItemList, pItem);
+    assert(!m_CreatedItemList.insert(pItem).second); // assert if item was already in the collection.
```
Comment (ccw808) [inline Client/core/Graphics/CRenderItemManager.cpp]: The standard library definition of assert might not actually compile the condition code: https://en.cppreference.com/w/cpp/error/assert. Although MTA uses it's own assert definition I still think it would be good practice to have required code outside of the assert. e.g.
```
bool bCheck = condition();
assert(bCheck);
```

File: Shared/sdk/CVector2D.h [hunk 15]
```diff
@@ -20,126 +20,121 @@
 class CVector2D
 {
 public:
+    float fX = 0.0f, fY = 0.0f;
```
Comment (qaisjp) [inline Shared/sdk/CVector2D.h:23]: separate lines please.

File: Client/core/Graphics/CRenderItemManager.cpp [hunk 16]
```diff
@@ -451,11 +449,10 @@ void CRenderItemManager::UpdateBackBufferCopySize()
     m_bBackBufferCopyMaybeNeedsResize = false;
 
     // Set what the max size requirement is for the back buffer copy
-    uint uiSizeX = 0;
-    uint uiSizeY = 0;
```
Comment (qaisjp) [inline Client/core/Graphics/CRenderItemManager.cpp]: Unnecessary changes

File: Client/core/Graphics/CRenderItemManager.cpp [hunk 17]
```diff
@@ -1040,9 +1044,7 @@ void CRenderItemManager::PreDrawWorld()
     IDirect3DTexture9*& pReadableDepthBuffer = g_pDeviceState->MainSceneState.DepthBuffer;
 
     // Determine what is needed
-    bool bRequireDepthBuffer = false;
-    if (!m_ShadersUsingDepthBuffer.empty() && m_depthBufferFormat != RFORMAT_UNKNOWN)
-        bRequireDepthBuffer = true;
```
Comment (qaisjp) [inline Client/core/Graphics/CRenderItemManager.cpp]: Older code is clearer.

File: Shared/sdk/CVector4D.h [hunk 18]
```diff
@@ -20,48 +20,32 @@
...
 {
 public:
-    CVector4D()
-    {
-        fX = 0;
-        fY = 0;
-        fZ = 0;
-        fW = 0;
-    }
+    float fX = 0.0f, fY = 0.0f, fZ = 0.0f, fW = 0.0f;
```
Comment (qaisjp) [inline Shared/sdk/CVector4D.h]: Keep this on separate lines as before

File: Client/core/Graphics/CRenderItemManager.cpp (hunk 14 above)
Comment (Pirulax) [reply Client/core/Graphics/CRenderItemManager.cpp]: Okay, no problem.

File: Client/core/Graphics/CGraphics.cpp (hunk 11 above)
Comment (Pirulax) [reply Client/core/Graphics/CGraphics.cpp]: The compiler optimized it away, I knew it, i just thought the new way would be cleaner.
But okay, no problem.

File: Client/core/Graphics/CGraphics.cpp (hunk 8 above)
Comment (Pirulax) [reply Client/core/Graphics/CGraphics.cpp:495]: well, just wanted to indicate neither of the 2 bools are used later. Putting the in a scope indicates that.

File: Client/core/Graphics/CGraphics.cpp (hunk 7 above)
Comment (Pirulax) [reply Client/core/Graphics/CGraphics.cpp]: I wanted to make it constexpr, for that you need static. But, I coudnt, so i left the static there. I'll remove it.

Comment (qaisjp) [reply Client/core/Graphics/CGraphics.cpp]: > I wanted to make it constexpr, for that you need static.

Nice, I didn't know that!

Comment (Pirulax) [reply Client/core/Graphics/CGraphics.cpp]: Nah, nvm, im stupid.

File: Client/core/Graphics/CGraphics.cpp (hunk 11 above)
Comment (Pirulax) [reply Client/core/Graphics/CGraphics.cpp]: im just so stuck with this one liner..
```cpp
while(wszText[textLength - 1 - spaceCount++] != L' ');
```
this is my new, totally not confusing method. It has the same efficiency as the others. But alright, I'll leave the current for loop version. I agree its kinda cleaner.

File: Client/core/Graphics/CGraphics.cpp (hunk 8 above)
Comment (Pirulax) [reply Client/core/Graphics/CGraphics.cpp:495]: Resolved?

**Comment by qaisjp** [comment]
This PR is ~~getting~~ too big + disorganised.

**Comment by Pirulax** [comment]
I agree. Ill open separate PRs.

File: Shared/sdk/CVector2D.h [hunk 19]
```diff
@@ -20,128 +20,122 @@
...
+    constexpr CVector2D operator*(const CVector2D& vecRight) const noexcept { return CVector2D(fX * vecRight.fX, fY * vecRight.fY); }
 
-    CVector2D operator/(const CVector2D& vecRight) const { return CVector2D(fX / vecRight.fX, fY / vecRight.fY); }
+    constexpr CVector2D operator/(const CVector2D& vecRight) const noexcept { return CVector2D(fX / vecRight.fX, fY / vecRight.fY); }
 
-    void operator+=(float fRight)
+    constexpr void operator+=(float fRight) noexcept
     {
         fX += fRight;
         fY += fRight;
```
Comment (ghost) [inline Shared/sdk/CVector2D.h:93]: I don't think `noexcept` is needed here. The CVector classes don't throw any exceptions. Also, I feel like `constexpr` is not useful here since the result is usually not known at compile time. The compiler is smart enough to know when to inline a function. This is just making the code hard to read.

Comment (Pirulax) [reply Shared/sdk/CVector2D.h:93]: please comment to #1478, and lets discuss it there.
tbh I agree, but nobody reads CVector code anyways. Constexpr is needed(at least for the constructor) in `CGraphics` refactor(somewhere in the end, the sphere thing)

**Comment by Pirulax** [comment]
As noted above, this is getting disorganized. I'll close it, and do separate PRs for the ease of testing, and merging.
